### PR TITLE
Fix compatibility issue around fuse_custom_io->clone_fd

### PR DESCRIPTION
**Pull request by legezywzh**
Fixes: 73cd124d0408 ("Add clone_fd to custom IO (#927)")

### CODE REVIEW

File: include/fuse_lowlevel.h [hunk 1]
```diff
@@ -2119,8 +2130,17 @@ fuse_session_new(struct fuse_args *args,
...
+#if (defined(LIBFUSE_BUILT_WITH_VERSIONED_SYMBOLS))
+/* Do not call this directly, but only through fuse_session_custom_io() */
+int _fuse_session_custom_io(struct fuse_session *se,
+			const struct fuse_custom_io *io,
+			size_t op_size, int fd);
+#else
+#define _fuse_session_custom_io(se, io, op_size, fd)	\
+	_fuse_session_custom_io_317(se, io, op_size, fd)
+#endif
+#define fuse_session_custom_io(se, io, fd)	\
```
Comment (bsbernd) [inline include/fuse_lowlevel.h:2142]: What is the purpose of this macro?  I had introduced the inlined function `fuse_session_new` which calls into `_fuse_session_new` to encode in the libfuse version an application was compiled against. _Maybe_ I should also do that for fuse_session_custom_io(). But right now this macro here does not bring an advantage?

**Comment by bsbernd** [comment]
@legezywzh Thanks for working on this, looks good to me, just a comment about the "_" name

File: include/fuse_lowlevel.h (hunk 1 above)
Comment (legezywzh) [reply include/fuse_lowlevel.h:2142]: > What is the purpose of this macro? I had introduced the inlined function `fuse_session_new` which calls into `_fuse_session_new` to encode in the libfuse version an application was compiled against. _Maybe_ I should also do that for fuse_session_custom_io(). But right now this macro here does not bring an advantage?

Do you mean change to below:
```
@@ -2119,8 +2130,13 @@ fuse_session_new(struct fuse_args *args,
  * @return -errno  if failed to allocate memory to store `io`
  *
  **/ 
+#if (defined(LIBFUSE_BUILT_WITH_VERSIONED_SYMBOLS))
 int fuse_session_custom_io(struct fuse_session *se,
                   const struct fuse_custom_io *io, int fd);
+#else
+#define fuse_session_custom_io(se, io, fd) \
+   _fuse_session_custom_io_317(se, io, sizeof(*io), fd)
+#endif
```

But original fuse_session_custom_io() API doesn't support to pass sizeof(fuse_custom_io ) info, so for LIBFUSE_BUILT_WITH_VERSIONED_SYMBOLS branch, it won't work.

Comment (bsbernd) [reply include/fuse_lowlevel.h:2142]: I only see
```
#if  (defined(LIBFUSE_BUILT_WITH_VERSIONED_SYMBOLS))
int _fuse_session_custom_io(struct fuse_session *se,...
...
#endif
#define fuse_session_custom_io(se, io, fd) _fuse_session_custom_io(se, io, sizeof(*io), fd)
```

I.e. redirection of `fuse_session_custom_io` to the underscored function

Comment (legezywzh) [reply include/fuse_lowlevel.h:2142]: Oh, I think now I got what you mean.
Indeed we can change original fuse_session_custom_io() API to add sizeof(fuse_custom_io ) info for LIBFUSE_BUILT_WITH_VERSIONED_SYMBOLS branch, right?

Comment (legezywzh) [reply include/fuse_lowlevel.h:2142]: Hmm, seems that I was wrong, API must not be changed, Could you please give some code hints.

**Comment by bsbernd** [comment]
Ah right, what about making it an inlined function instead of a macro?

```
int fuse_session_custom_io_317(struct fuse_session *se, const struct fuse_custom_io *io, size_t op_size, int fd);
#if FUSE_USE_VERSION < FUSE_MAKE_VERSION(3, 17)
static inline int fuse_session_custom_io(struct fuse_session *se, const struct fuse_custom_io *io, size_t op_size, int fd)
{
     return fuse_session_custom_io_317(se, io, op_size, fd);
}
#else
static inline int fuse_session_custom_io(struct fuse_session *se, const struct fuse_custom_io *io, int fd)
{
   return fuse_session_custom_io_317(se, io, offsetof(struct fuse_custom_io, clone_fd), fd);
}
#endif

```

There is also no underscore "_" and version check is "<" instead of "<=" as 3.17 is not released yet.

**Comment by bsbernd** [comment]
(personally I prefer inlined functions over macros, though just my personal opinion)

**Comment by legezywzh** [comment]
> Ah right, what about making it an inlined function instead of a macro?
> 
> ```
> int fuse_session_custom_io_317(struct fuse_session *se, const struct fuse_custom_io *io, size_t op_size, int fd);
> #if FUSE_USE_VERSION < FUSE_MAKE_VERSION(3, 17)
> static inline int fuse_session_custom_io(struct fuse_session *se, const struct fuse_custom_io *io, size_t op_size, int fd)
> {
>      return fuse_session_custom_io_317(se, io, op_size, fd);
> }
> #else
> static inline int fuse_session_custom_io(struct fuse_session *se, const struct fuse_custom_io *io, int fd)
> {
>    return fuse_session_custom_io_317(se, io, offsetof(struct fuse_custom_io, clone_fd), fd);
> }
> #endif
> ```
> 
> There is also no underscore "_" and version check is "<" instead of "<=" as 3.17 is not released yet.


then there'll be compile error:
../lib/compat.c:59:5: error: redefinition of ‘fuse_session_custom_io’
   59 | int fuse_session_custom_io(struct fuse_session *se,
      |     ^~~~~~~~~~~~~~~~~~~~~~

See patch codes about compat.c

**Comment by legezywzh** [comment]
> Ah right, what about making it an inlined function instead of a macro?
> 
> ```
> int fuse_session_custom_io_317(struct fuse_session *se, const struct fuse_custom_io *io, size_t op_size, int fd);
> #if FUSE_USE_VERSION < FUSE_MAKE_VERSION(3, 17)
> static inline int fuse_session_custom_io(struct fuse_session *se, const struct fuse_custom_io *io, size_t op_size, int fd)
> {
>      return fuse_session_custom_io_317(se, io, op_size, fd);
> }
> #else
> static inline int fuse_session_custom_io(struct fuse_session *se, const struct fuse_custom_io *io, int fd)
> {
>    return fuse_session_custom_io_317(se, io, offsetof(struct fuse_custom_io, clone_fd), fd);
> }
> #endif
> ```
> 
> There is also no underscore "_" and version check is "<" instead of "<=" as 3.17 is not released yet.

Or do you mean that using this code snippet, we don't need to do versioned symbols check about fuse_session_custom_io()?

**Comment by bsbernd** [comment]
We need versioned symbols for ABI, but for API I thought this code snippet would do. 

Regarding compat.c, this PR should fix your issue
https://github.com/libfuse/libfuse/pull/954

**Comment by bsbernd** [comment]
I merged PR #954, you shouldn't get the compilation error after rebase.

**Comment by legezywzh** [comment]
@bsbernd I have updated one new version, thanks for for your patience.

And I have one personal question, in the development of libfuse, API can be changed in the new libfuse version, right? It only requires that old ABI is still provided, and userspace filesystems codes need to adapt new API and re-compile.
Do I understand correctly?

**Comment by bsbernd** [comment]
Sorry I didn't manage to look into this today (actually already past midnight for me now), will respond after getting some sleep.

**Comment by legezywzh** [comment]
> Sorry I didn't manage to look into this today (actually already past midnight for me now), will respond after getting some sleep.


Really never mind, just take your time.

**Review by bsbernd — APPROVED** [review]

**Comment by bsbernd** [comment]
@legezywzh Thank you, also for the updates! LGTM, merged.

**Pull request merged**